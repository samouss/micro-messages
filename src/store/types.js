// @flow

import type { Store as ReduxStore, Dispatch as ReduxDispatch } from 'redux';
import type { MessageState, MessageAction } from '../modules/messages/types';

export type State =
  & MessageState;

export type Action =
  | MessageAction;

export type Store = ReduxStore<State, Action>;

export type GetState = () => State;

export type Thunk<A> = ((Dispatch, GetState) => Promise<void> | void) => A;

export type Dispatch =
  & ReduxDispatch<Action>
  & Thunk<Action>;
