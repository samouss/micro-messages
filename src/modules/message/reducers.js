// @flow

import { combineReducers } from 'redux';
import * as ACTION_TYPES from './actionTypes';
import type { Action } from 'store/types';
import type { MessageByIdState, MessageIdsState } from './types';

const messageByIdInitialState: MessageByIdState = {};
const messageById = (state = messageByIdInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return {
        ...state,
        ...action.messages.reduce((acc, message) => ({
          ...acc,
          [message.id]: message,
        }), {}),
      };
    }

    default: {
      return state;
    }
  }
};

const messageIdsInitialState: MessageIdsState = [];
const messageIds = (state = messageIdsInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return state.concat(action.messages.map(message => message.id));
    }

    default: {
      return state;
    }
  }
};

export default combineReducers({
  messageById,
  messageIds,
});
