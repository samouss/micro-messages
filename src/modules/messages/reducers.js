// @flow

import { combineReducers } from 'redux';
import * as ACTION_TYPES from './actionTypes';
import type { State, Action } from 'store/types';
import type { MessageId, ByIdState, IdsState } from './types';

const byIdInitialState: ByIdState = {};
const byId = (state = byIdInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return {
        ...state,
        ...action.messages.reduce((acc, message) => ({
          ...acc,
          [message.id]: message,
        }), {}),
      };
    }

    default: {
      return state;
    }
  }
};

const idsInitialState: IdsState = [];
const ids = (state = idsInitialState, action: Action) => {
  switch (action.type) {
    case ACTION_TYPES.FETCH_MESSAGES_SUCCESS: {
      return state.concat(action.messages.map(message => message.id));
    }

    default: {
      return state;
    }
  }
};

export default combineReducers({
  byId,
  ids,
});

const getState = (state: State) => state.messages;
const getMessageIds = (state: State) => getState(state).ids;

export const getMessage = (state: State, id: MessageId) => {
  return getState(state).byId[id];
};

export const getMessages = (state: State) => {
  return getMessageIds(state).map(id => getMessage(state, id));
};
