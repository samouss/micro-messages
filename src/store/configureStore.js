// @flow

import { createStore, applyMiddleware } from 'redux';
import { createLogger } from 'redux-logger';
import thunk from 'redux-thunk';
import reducer from './rootReducer';

import type { Store } from './types';

const configureStore = (): Store => {
  const middlewares = [thunk];

  if (process.env.NODE_ENV !== 'production') {
    middlewares.push(createLogger());
  }

  const store = createStore(
    reducer,
    applyMiddleware(...middlewares),
  );

  return store;
};

export default configureStore;
