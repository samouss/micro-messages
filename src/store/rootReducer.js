// @flow

import { combineReducers } from 'redux';
import messages from 'modules/messages/reducers';

export default combineReducers({
  messages,
});
