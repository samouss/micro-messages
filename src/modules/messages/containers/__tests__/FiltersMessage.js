// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { createMockStore } from 'test/store';
import { getVisibilityFilter } from '../../reducers';
import Connect, { FiltersMessage } from '../FiltersMessage';

import type { Props } from '../FiltersMessage';

jest.mock('../../reducers', () => ({
  getVisibilityFilter: jest.fn(state => state),
}));

jest.mock('../../actions', () => ({
  changeVisibilityFilter: jest.fn(filter => ({
    type: 'CHANGE_VISIBILITY_FILTER',
    filter,
  })),
}));

describe('<FiltersMessage />', () => {
  const createContext = state => ({
    store: createMockStore(state),
  });

  const defaultProps: Props = {
    visibility: 'all',
    onChangeVisibilityFilter: () => {},
  };

  it('expect to render the component correctly', () => {
    const props: Props = {
      ...defaultProps,
    };

    const component = shallow(
      <FiltersMessage
        {...props}
      />,
    );

    expect(component).toMatchSnapshot();
  });

  it('expect to render the component correctly', () => {
    const props: Props = {
      ...defaultProps,
      onChangeVisibilityFilter: jest.fn(),
    };

    const component = shallow(
      <FiltersMessage
        {...props}
      />,
    );

    component
      .find('FiltersMessageForm')
      .simulate('change');

    expect(props.onChangeVisibilityFilter).toHaveBeenCalled();
  });

  describe('mapStateToProps', () => {
    it('expect to call getVisibilityFilter', () => {
      const state = { visibilityFilter: 'all' };
      const context = createContext(state);

      shallow(
        <Connect />,
        { context },
      );

      expect(getVisibilityFilter).toHaveBeenCalledWith(state);
    });
  });

  describe('mapDispatchToProps', () => {
    it('expect to bind changeVisibilityFilter', () => {
      const filter = 'public';
      const context = createContext();

      const component = shallow(
        <Connect />,
        { context },
      );

      // $FlowFixMe
      component.props().onChangeVisibilityFilter(filter);

      expect(context.store.dispatch).toHaveBeenCalledWith({
        type: 'CHANGE_VISIBILITY_FILTER',
        filter,
      });
    });
  });
});
