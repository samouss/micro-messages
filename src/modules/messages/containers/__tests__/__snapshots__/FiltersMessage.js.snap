// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FiltersMessage /> expect to render the component correctly 1`] = `
<FiltersMessageForm
  onChange={[Function]}
  visibility="all"
/>
`;
